### ISSUE TITLE
Missing announcements of NVDA on Java Swing JTree via Java Access Bridge

### ISSUE THREAD

**Issue by anOolamr** [nvaccess/nvda#13623]
When navigating with the keyboard in a JTree, NVDA does not announce the structure and relationship information of the JTree items although this information is transmitted via the Java Access Bridge. In addition, NVDA does not announce the expand/collapse action on a JTree node although this state change is transmitted via the Java Access Bridge. 
Other screen readers announce the structure and relationship information as well as the expand/collapse state change.

The Web Content Accessibility Guidelines (WCAG) and the European Norm 301 549 require the availability of the mentioned information. See for example the following requirements:
WCAG 2.1, 1.3.1, Level A, Info and Relationships
https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html
WCAG 2.1, 4.1.2, Level A, Name, Role, Value
https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html
EN 301 549 V3.1.1 (2019-11) 9.4.1.2, 10.4.1.2, and 11.4.1.2
EN 301 549 V3.1.1 (2019-11) 9.1.3.1, 10.1.3.1, 11.1.3.1
https://www.etsi.org/deliver/etsi_en/301500_301599/301549/03.01.01_60/en_301549v030101p.pdf

There are county-specific regulations which request the fulfillment of the WCAG 2.1 or the EN 301 549 requirements.

### Steps to reproduce:
Prerequisites:
You have installed a Java Runtime Environment, e.g. JDK/JRE version 17 or 18.
You have downloaded the SwingSet2 Demonstration Program from: https://docs.oracle.com/javase/tutorial/uiswing/lookandfeel/plaf.html#swingset2

Steps, actual behavior and expected behavior:
1. Start NVDA (if not already running).
2. After starting NVDA, launch SwingSet2.jar with the Java JRE.
-> The SwingSet2 application window appears.

3. Navigate with arrow left key to the "JTree demo" radio button and press spacebar to select it. (The NVDA announcement is incomplete, but this issue is not discussed here.)
4. Press Tab key twice to navigate to the JTree control.
-> NVDA announces the "Tree" role of the JTree control.
5. Press arrow down key once to select the root node.
-> NVDA annouunces the name "Music" and some state information ("selected, expanded, editable") of the root node.
-> Missing announcements: hierarchy level of the node (level 0) and the number of its child items ("3 items").
6. Press arrow down key once to select the first child node.
-> NVDA annouunces the name "Classical" and some state information ("selected, collapsed, editable") of the node.
-> Missing announcements: hierarchy level of the node ("level 1") and its relationship to the other child items ("1 of 3").
7. Press arrow right to expand the "Classical" tree node.
-> NVDA announces nothing. NVDA should announce the name and the state change ("expanded") of the node and the number of its child items ("3 items").
8. Press arrow left to collapse the "Classical" tree node.
-> NVDA announces nothing. NVDA should announce the name and the state change ("collapsed") of the node and its relationship to the other items on the same level.
9. Press arrow right to expand the "Classical" tree node again and then press arrow down to navigate to its first child "Beethoven".
-> NVDA announces the name and the state information of the tree node.
-> Missing announcements: hierarchy level of the node and the number of its child items.
10. Press Shift+Tab command to navigate backwards out of the JTree. Then press Tab key to navigate back to the JTree.
-> NVDA announces the "Tree" role, the name and the state information of the "Beethoven" tree item.
-> Missing announcements: the hierarchy level of the "Beethoven" item and its relationship to the other child items ("1 of 3").
11. Press arrow right to expand the "Beethoven" tree node and then press arrow down to navigate to its first child "Concertos".
12. Press arrow right to expand the "Concertos" tree node and then press arrow down to navigate to its first child "No. 1 - C".
-> NVDA announces the name and the state information of the tree node.
-> Missing announcements: hierarchy level of the node, that it is a leaf (it has no child items), and its relationship to the other items on the same level.
13. Press arrow up to navigate back to the parent node "Concertos".
-> NVDA announces the name and the state information of the tree node.
-> Missing announcements: hierarchy level of the node and the number of its child items. 

#### NVDA installed/portable/running from source:
installed

#### NVDA version:
Version: 2021.3.5

#### Windows version:
Version	10.0.19042 Build 19042

#### Name and version of other software in use when reproducing the issue:
java version "18.0.1" 2022-04-19
Java(TM) SE Runtime Environment (build 18.0.1+10-24)
Java HotSpot(TM) 64-Bit Server VM (build 18.0.1+10-24, mixed mode, sharing)

#### Other information about your system:
Java Access Bridge is enabled.

#### Does the issue still occur after restarting your computer?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes. Same behavior.

#### If add-ons are disabled, is your problem still occurring?
Not applicable.

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not applicable.

**Comment by mohammad-suliman** [comment]
The issue of not reporting expanded / collapsed states is documented very Thurley in #5812. The bottom line that Java Access bridge has some design flow which makes resolving the issue somewhat challenging. Not sure what NV Access will do regarding this today, this ticket was opened in 2016 when a serious work on Android Studio was going on. The other issue of not reporting the node level is very strange, #5766 was closed with pull request which fixes this which makes me wonder why this has regressed with the current version. Regarding the issue of NVDA not reporting the relation to other nodes, e.g. item 1 of 3, I am not sure I can add something beneficial because I don't remember this particular issue.

**Comment by anOolamr** [comment]
Thanks for your feedback! Yes, I saw this 6 year old issue and I decided to create a new one with all issues regarding JTree in one bug report.

Java's jaccessinspector.exe (part of jre, in \bin folder) is a very helpful tool to monitor the Java Access Bridge API from "outside" on Windows OS (like a screen reader). When I expand the "Classical" node in JTree of SwingSet2, jaccessinspector reports a state change event with the following AccessibleContext information:

State changed event: old = collapsed; new = expanded

Version Information:
    Java virtual machine version: 18.0.1
    Access Bridge Java class version: 18.0.1
    Access Bridge Java DLL version: 18.0.1
    Access Bridge Windows DLL version: 18.0.1

AccessibleContext information:
    Name:  Classical
    Virtual Name:  Classical
    Description:  
    Role:  label
    Role in en_US locale:  label
    States:  aktiviert,fokussierbar,sichtbar,angezeigt,ausgewählt,aktiv,erweiterbar,eingeblendet,bearbeitbar
    States in en_US locale:  enabled,focusable,visible,showing,selected,active,expandable,expanded,editable
    Index in parent:  0
    Children count:  3
    Bounding rectangle:  [28, 318, 170, 354]
    Top-level window name:  SwingSet2
    Top-level window role:  frame
    Parent name:  Music
    Parent virtual name:  Music
    Parent role:  label
    Visible descendents count:  3
    Descendent 0 name:  Beethoven
    Descendent 0 role:  label
    Descendent 1 name:  Brahms
    Descendent 1 role:  label
    Descendent 2 name:  Mozart
    Descendent 2 role:  label

Accessible Selection information:
    Selection count:  0

AccessibleActions info:
    Number of actions:  1

**Comment by mohammad-suliman** [comment]
@anOolamr I believe you closed the issue by mistake. Yes I see. The issues I referenced are old, some of them is supposed to be fixed, and some still have no fix. I think that the mentioned issues need to get NV Access attention. Java Access bridge has some bugs that can be fixed on NVDA side, and I think that this needs to happen as soon as possible.

**Comment by anOolamr** [comment]
I reopened the issue, thanks a lot for your hint! I accidentally used the wrong button to add my comment. :-(

**Comment by mwhapples** [comment]
I have tried replicating this bug, I can only observe some of the issues mentioned.
* Item level: This seems to be correctly spoken when the tree gains focus or when moving between levels. NVDA refers to the root in the demo application as level 1, the next level as level2, etc, where as in the original issue report it was said expected would be level0 for the root item. This difference though I would not consider a bug.
* Child counts: I don't recall ever getting that with any other treeview in NVDA, so is it really expected for NVDA to speak child count?
* Relative position: NVDA does speak things like "1 of 3" and so on correctly.
* Expand/collapse announcement: NVDA does not speak this, NVDA speaks it for other treeviews. So I can replicate this part of the bug report.
* Reporting item as leaf node: NVDA tells me for non-leaf items the collapse/expanded state, leaf items have no collapsed/expanded state spoken, so leaf nodes seem to be distinguished. I do not recall NVDA ever announcing leaf item specifically in any other treeviews, so does not seem like less is being given for Java applications. Therefore I think NVA works as expected here.

I have tested this on two separate computers and get the same results on both machines. Does anyone else observe the regressions?

**Comment by anOolamr** [comment]
Thanks for your comment. As mentioned in #13848 I went through all of my NVDA settings again and found that the "Object Presentation" > "Report object position information" was deactivated. When I mark this checkbox, NVDA properly announces the hierarchy level of a tree node and its relationship to the other items on the same level. Sorry for any inconveniences.

Regarding your comment:

> Item level: This seems to be correctly spoken when the tree gains focus or when moving between levels. NVDA refers to the root in the demo application as level 1, the next level as level2, etc, where as in the original issue report it was said expected would be level0 for the root item. This difference though I would not consider a bug.

Right. Starting with "Level 1" for the root is fine with me.

> Child counts: I don't recall ever getting that with any other treeview in NVDA, so is it really expected for NVDA to speak child count?

I propose to announce the child count in addition when a node is expanded, for example: "expanded, 3 items".

> Relative position: NVDA does speak things like "1 of 3" and so on correctly.

Right, there is no issue.

> Expand/collapse announcement: NVDA does not speak this, NVDA speaks it for other treeviews. So I can replicate this part of the bug report.

Yes, both announcements are missing.

> Reporting item as leaf node: NVDA tells me for non-leaf items the collapse/expanded state, leaf items have no collapsed/expanded state spoken, so leaf nodes seem to be distinguished. I do not recall NVDA ever announcing leaf item specifically in any other treeviews, so does not seem like less is being given for Java applications. Therefore I think NVA works as expected here.

OK, this is not a bug. Nevertheless, I prefer an explicit information than relying on a missing state announcement,

**Comment by mwhapples** [comment]
I suggest for the part relating to number of child items and the explicit announcement of leaf nodes, creating new issues as these would be enhancements for all tree views in NVDA. Leave this issue open as there is still the issue relating to expanded/collapsed announcements which I would like to get fixed at some point although it is proving to be a slightly tricky bug to deal with.

**Comment by anOolamr** [comment]
- Bug reports #13891 and #13892 created.
- Title of this bug report changed from "Missing announcements of NVDA on Java Swing JTree via Java Access Bridge" to "NVDA does not announce the expand/collapse state change of an JTree node (Java Access Bridge API)".

**Comment by Adriani90** [comment]
Only steps 7 and 8 are reprotducible with NVDA 2024.4.1 and these are already covered in #5812. The "expanded" and "colapsed" message is not announced when expanding or colapsing tree objects with left and right arrow key. In Windows explorer this is properly announced.
cc: @dmitrii-drobotov is this something that needs to be exposed properly by the API?

@anOolamr you wrote:
> I propose to announce the child count in addition when a node is expanded, for example: "expanded, 3 items".

This is extra verbosity which is not needed. when you expand an item and press down arrow key, you already hear the count of the child items. The less verbosity we need to transmit the same information, the better it is for user's efficiency.

In step 12 you wrote:
> Missing announcements: hierarchy level of the node, that it is a leaf (it has no child items), and its relationship to the other items on the same level.

Level and count is announced when object presentation is enabled.
The fact that the property "colapsed" or "expanded" is not announced on the child element is already clear enough saying that there are no further child elements.

These being said I am closing in favor of the remaining #5812.
